scripts/sas-trading.js

```diff
@@ -1,2 +1,277 @@
+            type: Object,
+            default: {}
+        })
+        game.settings.register(this.ID, this.SETTINGS.CITIES, {
+            scope: 'world',
+            config: false,
+            type: Object,
+            default: []
+        })
+        // TODO: register a settings menu to fill stuff in
+    }
+
+    /**
+     * getSettings is a helper to retrieve a setting with this module's ID.
+     * @param {string} settingName 
+     * @returns {(SasGoods|SasBaseGoods|SasCities)} See SETTINGS for more info.
+     */
+    static getSetting(settingName) {
+        return game.settings.get(this.ID, settingName)
+    }
+
+    /**
+     * setSetting is a helper to store a setting with this module's ID.
+     * @param {string} settingName 
+     * @param {(SasGoods|SasBaseGoods|SasCities)} data See SETTINGS for typedefs
+     */
+    static async setSetting(settingName, data) {
+        return game.settings.set(this.ID, settingName, data)
+    }
+
+    static initialize() {
+        this.registerSettings()
+    }
+}
+
+class SasTradingGoodData {
+    /**
+     * Demand is an enum representing the different demand levels of trade goods.
+     * @see {SasDemand}
+     */
+    static Demand = Object.freeze({
+        HIGH: Symbol('high'),
+        AVG: Symbol('avg'),
+        LOW: Symbol('low')
+    })
+
+    /**
+     * Scarcity is an enum representing the different scarcity levels of trade
+     * goods.
+     * @see {SasScarcity}
+     */
+    static Scarcity = Object.freeze({
+        RARE: Symbol('rare'),
+        COMMON: Symbol('common'),
+        ABUNDANT: Symbol('abundant')
+    })
+
+    /**
+     * goodId generates an ID for the combination of good and city.
+     * @param {SasGoodName} goodName The name of the trade good
+     * @param {string} city The city for the trade good
+     * @returns {SasGoodId}
+     */
+    static goodId(goodName, city) {
+        return `${goodName}#${city}`
+    }
+
+    /**
+     * createGood adds a good to the DB indexed by the ID.
+     * 
+     * If good.id is left undefined, it will be generated by createGood.
+     * If a good with this ID already exists, it does nothing.
+     * @param {SasGood} good 
+     */
+    static async createGood(good) {
+        if (!good.id) {
+            good.id = this.goodId(good.name, good.city)
+        }
+        const goods = this.allGoods
+        if (goods.hasOwnProperty(good.id)) {
+            SasTrading.log(false, "good already exists:", good.id)
+            // TODO: Application Warning? Error?
+            return
+        }
+
+        goods[good.id] = good
+        await SasTrading.setSetting(SasTrading.SETTINGS.GOODS, goods)
+    }
+
+    /**
+     * updateGood updates an existing good by ID.
+     * 
+     * This merges the existing good with the provided good.
+     * If good.id is left undefined, it will be filled in.
+     * @param {SasGoodId} id 
+     * @param {SasGood} good 
+     */
+    static async updateGood(id, good) {
```

Comment:
Because `name` and `city` make up the good's ID, they can never be changed. Add checks for this, as well as the provided ID correctness if `name` and `city` are in the provided update

Reply:
Done